### PR TITLE
[MNG-7102] The child modules of excluded projects are now excluded

### PR DESCRIPTION
**Pull request by MartinKanters**
Following this checklist to help us incorporate your 
contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/MNG-7102) filed 
       for the change (usually before you start working on it).  Trivial changes like typos do not 
       require a JIRA issue.  Your pull request should address just this issue, without 
       pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[MNG-XXX] - Fixes bug in ApproximateQuantiles`,
       where you replace `MNG-XXX` with the appropriate JIRA issue. Best practice
       is to use the JIRA issue title in the pull request title and in the first line of the 
       commit message.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will 
       be performed on your pull request automatically.
 - [x] You have run the [Core IT][core-its] successfully.

If your pull request is about ~20 lines of code you don't need to sign an
[Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf) if you are unsure
please ask on the developers list.

To make clear that you license your contribution under 
the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
you have to acknowledge this by using the following check-box.

 - [x] I hereby declare this contribution to be licenced under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)

 - [x] In any other case, please file an [Apache Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

[core-its]: https://maven.apache.org/core-its/core-it-suite/

### CODE REVIEW

File: maven-core/src/test/java/org/apache/maven/graph/DefaultGraphBuilderTest.java [hunk 1]
```diff
@@ -147,6 +147,9 @@
                         .excludedProjects( MODULE_B )
                         .makeBehavior( REACTOR_MAKE_UPSTREAM )
                         .expectResult( PARENT_MODULE, MODULE_C, MODULE_A, MODULE_C_2 ),
+                scenario( "Excluding a project additionally excludes its children" )
```
Comment (rfscholte) [inline maven-core/src/test/java/org/apache/maven/graph/DefaultGraphBuilderTest.java:150]: LGTM, indeed what I would expect

**Comment by michael-o** [comment]
"child modules" not "children modules"

**Comment by michael-o** [comment]
I wonder what would happen if I have this:

root: mod_a, mod_b, mod_c
mod_c: mod_c_1, mod_c_2, mod_c_3
mod_c_2: mod_c_2_1, mod_c_2_2, mod_c_2_3
mod_c_2_2: mod_c_2_2_1, mod_c_2_2_2

`mvn -pl !mod_c2,mod_c_2_2`

I think this should be the equivalent of `mvn -pl mod_c_2_2`.

**Comment by MartinKanters** [comment]
> I wonder what would happen if I have this:
> 
> root: mod_a, mod_b, mod_c
> mod_c: mod_c_1, mod_c_2, mod_c_3
> mod_c_2: mod_c_2_1, mod_c_2_2, mod_c_2_3
> mod_c_2_2: mod_c_2_2_1, mod_c_2_2_2
> 
> `mvn -pl !mod_c2,mod_c_2_2`
> 
> I think this should be the equivalent of `mvn -pl mod_c_2_2`.

@michael-o Good point, I discussed this with @mthmulders and we both agree your desired outcome should be reality. 
It was not the case though, so we had to change the DefaultGraphBuilder a bit more, please see the last commit.

**Review by michael-o — APPROVED** [review]

**Comment by rfscholte** [comment]
> I wonder what would happen if I have this:
> 
> root: mod_a, mod_b, mod_c
> mod_c: mod_c_1, mod_c_2, mod_c_3
> mod_c_2: mod_c_2_1, mod_c_2_2, mod_c_2_3
> mod_c_2_2: mod_c_2_2_1, mod_c_2_2_2
> 
> `mvn -pl !mod_c2,mod_c_2_2`
> 
> I think this should be the equivalent of `mvn -pl mod_c_2_2`.

AFAIK Maven should first pick all selected modules (`mod_c_2_2`) and remove all excluded modules (`mod_c2` recursive). In that case no project would be selected.

**Comment by michael-o** [comment]
> > I wonder what would happen if I have this:
> > root: mod_a, mod_b, mod_c
> > mod_c: mod_c_1, mod_c_2, mod_c_3
> > mod_c_2: mod_c_2_1, mod_c_2_2, mod_c_2_3
> > mod_c_2_2: mod_c_2_2_1, mod_c_2_2_2
> > `mvn -pl !mod_c2,mod_c_2_2`
> > I think this should be the equivalent of `mvn -pl mod_c_2_2`.
> 
> AFAIK Maven should first pick all selected modules (`mod_c_2_2`) and remove all excluded modules (`mod_c2` recursive). In that case no project would be selected.

I am trying to understand why this should be the case....

**Comment by rfscholte** [comment]
`mvn -pl !mod_c2,mod_c_2_2` is the same as `mvn -pl !mod_c2,!mod_c_2_1,!mod_c_2_2,!mod_c_2_2_1,!mod_c_2_2_2,!mod_c_2_3,mod_c_2_2`
Order doesn't matter. Because `mod_c_2_2` is both selected and deselected, it is effectively not selected.

**Comment by MartinKanters** [comment]
I interpreted @michael-o 's suggestion as that the more specific project would win. 
But anyway, I guess if selection then exclusion is a hard rule, then that makes sense as well. It's easier to explain to people as well. 
So, are you on board with this, @michael-o ? Then we'll go ahead and rework this PR.

**Comment by michael-o** [comment]
I agree with @rfscholte that order should not matter because this would make it much more complex. According to his explanation exclusion should win? If so, I am fine with that. We do the same for resource matching, exclusion wins.

**Comment by MartinKanters** [comment]
@michael-o @rfscholte Restored the original PR, so now it implements the rule `select, then exclude` again. Since this PR is required for other PRs and has been checked by the two of you and also by @mthmulders , I will go ahead and merge it.

**Comment by rfscholte** [comment]
I would like to see a test for `mvn -pl !mod_c2,mod_c_2_2` as well, as this was part of the discussion and hasn't been covered. Result should be: no project selected.

**Comment by mthmulders** [comment]
> I would like to see a test for `mvn -pl !mod_c2,mod_c_2_2` as well, as this was part of the discussion and hasn't been covered. Result should be: no project selected.

Good idea! We checked it manually and it failed with the expected "no projects selected" message. We will make an automated test in a separate PR, after #446 has been merged. That PR introduces a mechanism in `DefaultGraphBuilderTest` to expect a failed build, and "no project selected" will fail the build.

**Comment by michael-o** [comment]
I'd like to run the new test when ready.

**Comment by jira-importer** [comment]
Resolve #8350

**Pull request merged**